### PR TITLE
Psk Changes

### PR DESCRIPTION
**Pull request by 89jd**
Use config flow. Added new methods. Create app entitys so they can be launched from dashboard. Allow to use CEC rather than inputs

### CODE REVIEW

**Comment by gerard33** [comment]
Thanks for this impressive PR. Did not have time to test and review this yet.
Will try to do that asap.

**Comment by 89jd** [comment]
No worries. Been using it myself for past month or so. 

Was a good way of working out how home assistant works under the hood

**Comment by gerard33** [comment]
I have tested this and the config_flow works. Nice piece of work 👍

I did notice though that there are also some changes to the other parts of the code which give some errors in the HA log.
And I think it also needs an OptionFlow to be added. That way people can make changes in the configuration from the intergrations page without deleting and re-adding the integration.

Not sure yet how to proceed with this PR yet. Migrating to the config_flow will take time from my side to test this further and it's a breaking change for the current users. And time is the blocking item here 😬
I can make merge this in a separate branch and make this available as a beta version in HACS.

What do you think?

**Comment by 89jd** [comment]
Great. For the options flow, sure can work on that, I had it in at one point, but took it out for reasons I can't remember now :).

For the updates to the underlying PSK library, what is the best thing to do with that? Should we push it to your repo and have that as the base? The manifest currently points to my master repo directly.

I think it should be fine remapping it to be able to use yaml again though, did I actually explicitly take it out (been a little while :smile:)?

One other thing I remember doin, but can't remember why now, was I had to comment out the user labels stuff.

I'd be happy for it to go into a beta thing though.

f  you could do me a favor and raise them as issues, and I will get to them when I next get chance, probably in next couple of weeks

File: custom_components/braviatv_psk/manifest.json [hunk 1]
```diff
@@ -2,7 +2,8 @@
   "domain": "braviatv_psk",
   "name": "Bravia TV PSK",
   "documentation": "https://github.com/custom-components/media_player.braviatv_psk",
-  "requirements": ["pySonyBraviaPSK==0.2.0"],
+  "requirements": ["git+https://github.com/JackED42/braviarc.git@master#pySonyBraviaPSK==2.0"],
```
Comment (rchl) [inline custom_components/braviatv_psk/manifest.json:5]: I could help with testing this but I wonder if you have created a PR against pySonyBraviaPSK with your changes also?

Comment (89jd) [reply custom_components/braviatv_psk/manifest.json:5]: Yeh I raised pr, didn't hear anything from the developer...

Comment (rchl) [reply custom_components/braviatv_psk/manifest.json:5]: Did you raise it in the right place? I can't see anything in https://github.com/gerard33/sony_bravia_psk/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc

**Comment by 89jd** [comment]
https://github.com/aparraga/braviarc/pull/34

I think I used this one. For now I can't remember why I did that now. I think it semed more up to date

**Comment by rchl** [comment]
> [aparraga/braviarc#34](https://github.com/aparraga/braviarc/pull/34)
> 
> I think I used this one. For now I can't remember why I did that now. I think it semed more up to date

I suppose there is a higher chance of changes being integrated when submitting PR against the @gerard33 's fork which is the one that this integration is using anyway.

**Comment by gerard33** [comment]
You should indeed make the PR against my fork. As rchl already mentioned that is the one used in this integration.
I can then merge that and release the package as a beta on PyPi.
Be aware that there are some differences between the packages, as I added some code and afterwards there is code added to the aparraga version as well.

**Comment by sir-max** [comment]
Hi, I have tested the plugin with all the changes in this PR. Unfortunately I do not get around this error after setting up the plugin via gui:

`Logger: homeassistant.components.media_player
Source: custom_components/braviatv_psk/media_player.py:219
Integration: Mediaplayer (documentation, issues)
First occurred: 23:32:38 (1 occurrences)
Last logged: 23:32:38

Error while setting up braviatv_psk platform for media_player
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.8/site-packages/homeassistant/helpers/entity_platform.py", line 199, in _async_setup_platform
    await asyncio.shield(task)
  File "/home/homeassistant/.homeassistant/custom_components/braviatv_psk/media_player.py", line 219, in async_setup_entry
    unique_id = sys_info['serial']
KeyError: 'serial'`

Does anyone know why this happens? I could not find the problem here

**Comment by 89jd** [comment]
> You should indeed make the PR against my fork. As rchl already mentioned that is the one used in this integration.
> I can then merge that and release the package as a beta on PyPi.
> Be aware that there are some differences between the packages, as I added some code and afterwards there is code added to the aparraga version as well.

I'll get that updated for sure. It mixed both versions I think because they both had the functions I needed...

**Comment by 89jd** [comment]
> Hi, I have tested the addon with all the changes in this PR. Unfortunately I do not get around this error after setting up the plugin via gui:
> 
> ```
> Logger: homeassistant.components.media_player
> Source: custom_components/braviatv_psk/media_player.py:219
> Integration: Mediaplayer (documentation, issues)
> First occurred: 23:32:38 (1 occurrences)
> Last logged: 23:32:38
> 
> Error while setting up braviatv_psk platform for media_player
> Traceback (most recent call last):
>   File "/srv/homeassistant/lib/python3.8/site-packages/homeassistant/helpers/entity_platform.py", line 199, in _async_setup_platform
>     await asyncio.shield(task)
>   File "/home/homeassistant/.homeassistant/custom_components/braviatv_psk/media_player.py", line 219, in async_setup_entry
>     unique_id = sys_info['serial']
> KeyError: 'serial'
> ```
> 
> Does anyone know why this happens? I could not find the problem here

I'm afraid I will struggle to debug. In between houses so the TV is not setup currently. Looksike it isn't getting the serial number on request to the TV. Hard to say much more than that. I'd recommend looking a few lines up from the one it's crashing on, and maybe printing sys_info

**Comment by sir-max** [comment]
> > Hi, I have tested the addon with all the changes in this PR. Unfortunately I do not get around this error after setting up the plugin via gui:
> > ```
> > Logger: homeassistant.components.media_player
> > Source: custom_components/braviatv_psk/media_player.py:219
> > Integration: Mediaplayer (documentation, issues)
> > First occurred: 23:32:38 (1 occurrences)
> > Last logged: 23:32:38
> > 
> > Error while setting up braviatv_psk platform for media_player
> > Traceback (most recent call last):
> >   File "/srv/homeassistant/lib/python3.8/site-packages/homeassistant/helpers/entity_platform.py", line 199, in _async_setup_platform
> >     await asyncio.shield(task)
> >   File "/home/homeassistant/.homeassistant/custom_components/braviatv_psk/media_player.py", line 219, in async_setup_entry
> >     unique_id = sys_info['serial']
> > KeyError: 'serial'
> > ```
> > 
> > 
> > Does anyone know why this happens? I could not find the problem here
> 
> I'm afraid I will struggle to debug. In between houses so the TV is not setup currently. Looksike it isn't getting the serial number on request to the TV. Hard to say much more than that. I'd recommend looking a few lines up from the one it's crashing on, and maybe printing sys_info

I have managed to print the output:

```
cat sys_info.debug 
{'name': 'BRAVIA', 'model': 'KD-55XD7004', 'language': 'ger'}
```
It is indeed missing the serial here. Is it nessecary for the addon to work?

**Comment by 89jd** [comment]
It is not a must. It was just in case people had multiple, to differentiate them. Would be interesting to know why it doesn't return anything, I wonder if another function somewhere also returns a serial number...

**Comment by sir-max** [comment]
> It is not a must. It was just in case people had multiple, to differentiate them. Would be interesting to know why it doesn't return anything, I wonder if another function somewhere also returns a serial number...

Ok, now I just replaced the sys_info[serial] with a hardcoded name and that worked. Device shows up but now I got two new errors related to "BraviaRC":

Logger: custom_components.braviatv_psk.media_player
Source: custom_components/braviatv_psk/media_player.py:480
Integration: Bravia TV PSK (documentation)
First occurred: 1:07:49 (67 occurrences)
Last logged: 1:19:28

```
No data received from TV.
Traceback (most recent call last):
  File "/home/homeassistant/.homeassistant/custom_components/braviatv_psk/media_player.py", line 352, in async_update
    await self._refresh_channels()
  File "/home/homeassistant/.homeassistant/custom_components/braviatv_psk/media_player.py", line 480, in _refresh_channels
    self._label_list = await self.hass.async_add_executor_job(self._braviarc.get_current_external_input_status)
AttributeError: 'BraviaRC' object has no attribute 'get_current_external_input_status'
```
```
Logger: braviarc.braviarc
Source: /srv/homeassistant/lib/python3.8/site-packages/braviarc/braviarc.py:401
First occurred: 1:07:47 (82 occurrences)
Last logged: 1:19:27

Exception: 'NoneType' object has no attribute 'get'
```

**Comment by rchl** [comment]
You probably didn't test with the related changes to the PSK library (https://github.com/aparraga/braviarc/pull/34/files).

**Comment by sir-max** [comment]
> You probably didn't test with the related changes to the PSK library (https://github.com/aparraga/braviarc/pull/34/files).

You are right, I deployed them now. But no luck:

```
Logger: braviarc.braviarc
Source: /srv/homeassistant/lib/python3.8/site-packages/braviarc/braviarc.py:394
First occurred: 20:24:37 (26 occurrences)
Last logged: 20:31:21

Exception: 'NoneType' object has no attribute 'get'
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.8/site-packages/braviarc/braviarc.py", line 385, in load_app_list
    cookies = self._recreate_auth_cookie()
  File "/srv/homeassistant/lib/python3.8/site-packages/braviarc/braviarc.py", line 379, in _recreate_auth_cookie
    cookies.set("auth", self._cookies.get("auth"))
AttributeError: 'NoneType' object has no attribute 'get'
```

```
Logger: custom_components.braviatv_psk.media_player
Source: custom_components/braviatv_psk/media_player.py:338
Integration: Bravia TV PSK (documentation)
First occurred: 20:24:37 (2 occurrences)
Last logged: 20:31:07

No data received from TV.
Traceback (most recent call last):
  File "/home/homeassistant/.homeassistant/custom_components/braviatv_psk/media_player.py", line 338, in async_update
    for app in self._app_list:
TypeError: 'NoneType' object is not iterable
```

**Comment by gerard33** [comment]
@sir-max it's easier if you use this in `manifest.json`. Then you are sure that the correct version for the config_flow is used.
```
requirements": ["git+https://github.com/JackED42/braviarc.git@master#pySonyBraviaPSK==2.0"]
```

**Comment by rchl** [comment]
I'm working on adapting those changes to the latest code and fixing some issues while at it.

Could you answer those questions that I have that about your changes?
 1) Why have you changed the services to accept `entity_ids` instead of `entity_id`? If it's on purpose then the code wasn't adjusted to handle those and I'm not sure it's necessary to support multiple ids in one call.
 2) You've added an option "enable_cookies" that does this:
```py
            if self._enable_cookies and not self._braviarc.is_connected():
                await self.hass.async_add_executor_job(
                    self._braviarc.connect, None, 'hass', 'Home assistant'
                )
```
Is that even relevant to the PSK method of connecting to the TV? Since this code didn't exist before and everything was working for everyone, I'm not sure why it was added. Possibly just something that came from the original component?

**Comment by rchl** [comment]
3. Also I'm not sure I agree with creating the "app.*" entities for all installed apps. I think those are fairly useless as they can't be interacted with and the list of apps is also exposed in the attributes of the media player entity so for automation purposes those are already available.

**Comment by gerard33** [comment]
I am planning to close this PR also looking at #59.
@89jd are you okay with that?

**Comment by 89jd** [comment]
I made this a long time ago! Ha. Forgot what I did now. If the official one supports everything here. Happy for it to be closed

**Pull request closed without merging**